lcls_tools/common/data_analysis/bmad_modeling/bmad_modeling.py

```diff
@@ -0,0 +1,387 @@
+    # fudge = 1 + dF
+    complexGain = amplNp * np.exp(1j * phasNp)
+    complexGainFudged = complexGain + amplNp * dF
+    amplF = np.abs(complexGainFudged)
+    phasF = np.degrees(np.angle(complexGainFudged))
+    pvdata_fudge = dict()
+    pvdata_reg = dict()
+    for indx, pv in enumerate(devices):
+        pvdata_fudge[f"{pv}:PACTMEAN"] = phasF[indx]
+        pvdata_fudge[f"{pv}:AACTMEAN"] = amplF[indx]
+        pvdata_reg[f"{pv}:P{attr}"] = phasF[indx]
+        pvdata_reg[f"{pv}:A{attr}"] = amplF[indx]
+    data_map = {"cavities": mdl_obj.all_data_maps["cavities"]}
+    tao_cmds = get_tao_cmds(pvdata, data_map)
+    evaluate_tao(tao, tao_cmds)
+    region_e_tot = tao.ele_gen_attribs(f"END{region}B")["E_TOT"] / 1e9
+    print(region_e_tot)
+
+
+def update_energy_gain_cu(tao, pvdata, mdl_obj):
+    """
+    Updates Cu Linac energy gain profile based on bending magnets,
+    calculates a fudge and modifies model's cavity amplitudes and phases
+    """
+    expected_gain = get_expected_energy_gain(mdl_obj.region)
+    init_cmds = ["veto dat *", "veto var *"]
+    if mdl_obj.region == "L2":
+        tao.cmd(f"set dat BC1.energy[2]|meas = {expected_gain} ")
+        optimize_cmds = init_cmds + [
+            "use dat BC2.energy[1]",
+            "use var linac_fudge[2]"]
+    if mdl_obj.region == "L3":
+        tao.cmd(f"set dat L3[2]|meas ={expected_gain} ")
+        optimize_cmds = init_cmds + [
+            "use dat L3.energy[2]",
+            "use var linac_fudge[3]"]
+    tao.cmds(optimize_cmds)
+    r = tao.cmd("run")
+    print(r)
+
+
+def update_datum_meas(tao, datum, mdl_obj, useDesing=True):
+    """
+    Updates datum with EPICS or design twiss values
+    """
+    [d2_name, d1_name] = datum.split('.')
+    element = tao.data(d2_name, d1_name)['ele_name']
+    TWISS = ["beta_a", "alpha_a", "beta_b", "alpha_b"]
+    cmd = []
+    if useDesing:
+        ele_twiss = tao.ele_twiss(element, which="design")
+        for ii, p in enumerate(TWISS):
+            cmd.append(f"set dat {datum}[{ii + 1}]|meas = {ele_twiss[p]}")
+    else:
+        device = tao.ele_head(element)["alias"]
+        EPICS_ATTRIBUTE = ["BETA_X", "ALPHA_X", "BETA_Y", "ALPHA_Y"]
+        pvs = [device + ":" + twiss for twiss in EPICS_ATTRIBUTE]
+        if mdl_obj.data_source in ["ACT", "DES"]:
+            measured = get_live(pvs)
+        elif mdl_obj.data_source == "ARCHIVE":
+            measured = lcls_archiver_restore(pvs, mdl_obj.date_time)
+        for ii, pv in enumerate(pvs):
+            cmd.append(
+                f"set dat {datum}[{ii + 1}]|meas = {measured[pvs[ii]]}")
+    tao.cmds(cmd)
+    tao.cmd(f"show dat {datum}")
+
+
+class BmadModeling:
+    def __init__(self, beam_path, data_source):
+        self.beam_path: str = beam_path
+        self.data_source: str = data_source
+        self.beam_code = ['1' if beam_path == 'cu_hxr' else '2'][0]
+        self.date_time: str = '2024-03-24T17:10:00.000000-08:00'
```

Comment:
In general it's better to define class attributes in the class instead of the constructor. For instance:

```
class BmadModeling:
    beam_path: str
    data_source: str
    beam_code: str
    date_time: str
    data_maps: dict

    def __init__(self, beam_path, data_source):
        self.beam_path: str = beam_path
        self.data_source: str = data_source
        self.beam_code = ['1' if beam_path == 'cu_hxr' else '2'][0]
        self.date_time: str = '2024-03-24T17:10:00.000000-08:00'
```